.github/workflows/rust.yml

```diff
@@ -26,8 +26,54 @@ jobs:
 
   integration:
     runs-on: ubuntu-18.04
-    container:
-      image: docker://rust:1.37.0-buster
     steps:
     - uses: actions/checkout@v1
-    - run: make test-integration
+
+    # Forks run in GitHub actions, as they don't have access to secrets.
+    - name: Test (Fork)
+      if: github.event.pull_request.head.repo.fork
+      run: make test-integration
+
+    # Create a build image on a Linkerd build host.
+    - name: Setup (Origin)
+      if: '!github.event.pull_request.head.repo.fork'
+      env:
+        DOCKER_HOST: "ssh://linkerd-docker"
+      run: |
+        mkdir -p ~/.ssh
+        # Create an identity file and protect before writing contents to it.
+        touch ~/.ssh/id && chmod 600 ~/.ssh/id
+        echo "${{ secrets.DOCKER_PRIVATE_KEY }}" >~/.ssh/id
+        # Use well-known public keys for the host to prevent middlemen.
+        echo "${{ secrets.DOCKER_KNOWN_HOSTS }}" >~/.ssh/known_hosts
+        # Configure host with BatchMode to prevent idlelocking.
+        (
+          echo "Host linkerd-docker"
+          echo "    User github"
+          echo "    Hostname ${{ secrets.DOCKER_ADDRESS }}"
+          echo "    IdentityFile ~/.ssh/id"
+          echo "    BatchMode yes"
+          echo "    ServerAliveInterval 60"
+        ) >~/.ssh/config
+        # Build once with output, then build again to get the img ref (should
+        # be a noop).
+        docker build -f .github/actions/remote/Dockerfile .
+        docker build -qf .github/actions/remote/Dockerfile . >image.ref
+
+    # Use the previously built image to run tests.
+    - name: Test (Origin)
+      if: '!github.event.pull_request.head.repo.fork'
+      env:
+        DOCKER_HOST: "ssh://linkerd-docker"
+      run: docker run --rm $(cat image.ref) test-integration
+
+
+    # Try to delete the build image.
+    - name: Cleanup (Origin)
+      if: always() && !github.event.pull_request.head.repo.fork
```

Comment:
Why is `always()` needed? I would think the same condition would be here as the previous non-forked steps.

Reply:
As in: `if: '!github.event.pull_request.head.repo.fork'`

Reply:
always makes this run when a prior step has failed